### ISSUE TITLE
org.bouncycastle.asn1.BEROctetString found in input should only contain DEROctetString since 1.63

### ISSUE THREAD

**Issue by silver0815** [bcgit/bc-java#788]
Hi, 

for a e-mail i got a exception since 1.63 till 1.67b04

```
Caused by: java.lang.IllegalArgumentException: org.bouncycastle.asn1.BEROctetString found in input should only contain DEROctetString
	at org.bouncycastle.asn1.BEROctetString.toBytes(BEROctetString.java:48)
	at org.bouncycastle.asn1.BEROctetString.<init>(BEROctetString.java:104)
	at org.bouncycastle.asn1.BEROctetString.<init>(BEROctetString.java:78)
	at org.bouncycastle.asn1.ASN1OctetString.getInstance(ASN1OctetString.java:140)
	at org.bouncycastle.asn1.cms.EncryptedContentInfo.<init>(EncryptedContentInfo.java:55)
	at org.bouncycastle.asn1.cms.EncryptedContentInfo.getInstance(EncryptedContentInfo.java:82)
	at org.bouncycastle.asn1.cms.EnvelopedData.<init>(EnvelopedData.java:84)
	at org.bouncycastle.asn1.cms.EnvelopedData.getInstance(EnvelopedData.java:131)
	at org.bouncycastle.cms.CMSEnvelopedData.<init>(CMSEnvelopedData.java:75)
	... 3 more
```

if i try do read the e-mail:
`SMIMEEnveloped smimeEnveloped = new SMIMEEnveloped(msg);`

Maybe the problem is in the e-mail itself, but at the moment i try to find the errorplace.... Maybe some got the exception too and had a idea for me? 

I'm trying to set up a test case...
Suggestions and ideas for solving the problem are very welcome!

Thx. Martin

**Comment by silver0815** [comment]
Update:
I compare now 1.62 with 1.63 and I think the change is in:
org.bouncycastle.asn1.ASN1OctetString in getInstance method.

In 1.62 it works with this code.
```
    public static ASN1OctetString getInstance(
        ASN1TaggedObject    obj,
        boolean             explicit)
    {
        ASN1Primitive o = obj.getObject();

        if (explicit || o instanceof ASN1OctetString)
        {
            return getInstance(o);
        }
        else
        {
            return BEROctetString.fromSequence(ASN1Sequence.getInstance(o));
        }
    }
```

In 1.63 this methode has huge change.


The "taggedObject"	is from typ "BERTaggedObject". So in 1.63 I am in this code:
```
        /*
         * constructed object which appears to be explicitly tagged and it's really implicit means
         * we have to add the surrounding octet string.
         */
        if (taggedObject.isExplicit())
        {
            ASN1OctetString singleSegment = ASN1OctetString.getInstance(o);

            if (taggedObject instanceof BERTaggedObject)
            {
                return new BEROctetString(new ASN1OctetString[]{ singleSegment });
            }

            // TODO Should really be similar to the BERTaggedObject case above:
//            return new DLOctetString(new ASN1OctetString[]{ singleSegment });
            return (ASN1OctetString)new BEROctetString(new ASN1OctetString[]{ singleSegment }).toDLObject();
        }	
```

In "return new BEROctetString(new ASN1OctetString[]{ singleSegment });" the exception is thrown.
The "singleSegment" is from type "BEROctetString" and has this value "string with 1000 chars".

In BEROctetString is this method:
Here is a cast to "DEROctetString " but the value is a "BEROctetString".

```
  static private byte[] toBytes(
        ASN1OctetString[]  octs)
    {
        ByteArrayOutputStream bOut = new ByteArrayOutputStream();

        for (int i = 0; i != octs.length; i++)
        {
            try
            {
                DEROctetString o = (DEROctetString)octs[i];

                bOut.write(o.getOctets());
            }
            catch (ClassCastException e)
            {
                throw new IllegalArgumentException(octs[i].getClass().getName() + " found in input should only contain DEROctetString");
            }
            catch (IOException e)
            {
                throw new IllegalArgumentException("exception converting octets " + e.toString());
            }
        }

        return bOut.toByteArray();
    }

```

Sorry, I don't have that much idea, but this change (in ASN1OctetString) is causing the "bug".

**Comment by silver0815** [comment]
Update 2:

The problem is in ASN1OctetString.getInstance with this if statement:
```
if (taggedObject.isExplicit())
        {
            ASN1OctetString singleSegment = ASN1OctetString.getInstance(o);

            if (taggedObject instanceof BERTaggedObject)
            {
                return new BEROctetString(new ASN1OctetString[]{ singleSegment });
            }

            // TODO Should really be similar to the BERTaggedObject case above:
//            return new DLOctetString(new ASN1OctetString[]{ singleSegment });
            return (ASN1OctetString)new BEROctetString(new ASN1OctetString[]{ singleSegment }).toDLObject();
        }
```
My "BERTaggedObject" ist explicit because in ASN1StreamParser in " ASN1Primitive readTaggedObject(boolean constructed, int tag) throws IOException" this code is executed:
```
  if (_in instanceof IndefiniteLengthInputStream)
        {
            return v.size() == 1
                ?   new BERTaggedObject(true, tag, v.get(0))
                :   new BERTaggedObject(false, tag, BERFactory.createSequence(v));
        }
```
The size of v is 1, so explicit "true" is set.

And then the ASN1OctetString is trying to generate the new objects, 

Maybe this code:
```
 if (o instanceof ASN1OctetString)
        {
            ASN1OctetString s = (ASN1OctetString)o;

            if (taggedObject instanceof BERTaggedObject)
            {
                return s;
            }

            return (ASN1OctetString)s.toDLObject();
        }
```
in ASN1OctetString.getInstance should come before this " if (taggedObject.isExplicit())" ?

**Comment by silver0815** [comment]
Update 3:
If i change in ASN1OctetString the code from:

```
        /*
         * constructed object which appears to be explicitly tagged and it's really implicit means
         * we have to add the surrounding octet string.
         */
        if (taggedObject.isExplicit())
        {
            ASN1OctetString singleSegment = ASN1OctetString.getInstance(o);

            if (taggedObject instanceof BERTaggedObject)
            {
                return new BEROctetString(new ASN1OctetString[]{ singleSegment });
            }

            // TODO Should really be similar to the BERTaggedObject case above:
//            return new DLOctetString(new ASN1OctetString[]{ singleSegment });
            return (ASN1OctetString)new BEROctetString(new ASN1OctetString[]{ singleSegment }).toDLObject();
        }

        if (o instanceof ASN1OctetString)
        {
            ASN1OctetString s = (ASN1OctetString)o;

            if (taggedObject instanceof BERTaggedObject)
            {
                return s;
            }

            return (ASN1OctetString)s.toDLObject();
        }
```

to 

```
        if (o instanceof ASN1OctetString)
        {
        	ASN1OctetString s = (ASN1OctetString)o;
        	
        	if (taggedObject instanceof BERTaggedObject)
        	{
        		return s;
        	}
        	
        	return (ASN1OctetString)s.toDLObject();
        }
        /*
         * constructed object which appears to be explicitly tagged and it's really implicit means
         * we have to add the surrounding octet string.
         */
        if (taggedObject.isExplicit())
        {
            ASN1OctetString singleSegment = ASN1OctetString.getInstance(o);

            if (taggedObject instanceof BERTaggedObject)
            {
                return new BEROctetString(new ASN1OctetString[]{ singleSegment });
            }

            // TODO Should really be similar to the BERTaggedObject case above:
//            return new DLOctetString(new ASN1OctetString[]{ singleSegment });
            return (ASN1OctetString)new BEROctetString(new ASN1OctetString[]{ singleSegment }).toDLObject();
        }
```

works it again....

**Comment by dghgit** [comment]
I think the real issue is toBytes(), ASN.1 being flexible to the point of tear inducing there's no "issue" with nesting BER data. I'm not sure how that got there, but I've changed toBytes() just to treat things as ASN1OctetString and to not look specifically for DER.

**Comment by silver0815** [comment]
Thank you very much with the BC 1.67b07 it works again! Thank you very much!

**Comment by silver0815** [comment]
Fixed, testes in BC 1.67b07.
